### ISSUE TITLE
how to use QAT Engine speed more Encryption Algorithm

### ISSUE THREAD

**Issue by skomen** [intel/QAT_Engine#30]
I have installed QAT Engine，and The QAT engine now supports RSA, DSA, DH, AES-128-CBC-HMAC-SHA1, AES-256-CBC-HMAC-SHA256, AES-256-CBC-HMAC-SHA1, AES-256-CBC-HMAC-SHA256 , TLS1-PRF in my system，How and where can  I add support algorithms ?

**Comment by paulturx** [comment]
Hi,
the full list of Quickassist 'Features'  is given in the README.md file of the QAT engine:-
## Features

* Synchronous and Asynchronous Operation
* Asymmetric PKE Offload
  * RSA Support with PKCS1 Padding for Key Sizes 1024/2048/4096.
  * DH Support for Key Sizes 768/1024/1536/2048/3072/4096.
  * DSA Support for Key Sizes 160/1024, 224/2048, 256/2048, 256/3072.
  * ECDH Support for the following curves:
    * NIST Prime Curves: P-192/P-224/P-256/P-384/P-521.
    * NIST Binary Curves: B-163/B-233/B-283/B-409/B-571.
    * NIST Koblitz Curves: K-163/K-233/K-283/K-409/K-571.
  * ECDSA Support for the following curves:
    * NIST Prime Curves: P-192/P-224/P-256/P-384/P-521.
    * NIST Binary Curves: B-163/B-233/B-283/B-409/B-571.
    * NIST Koblitz Curves: K-163/K-233/K-283/K-409/K-571.
* Symmetric Chained Cipher Offload with pipelining capability:
  * AES128-CBC-HMAC-SHA1/AES256-CBC-HMAC-SHA1.
  * AES128-CBC-HMAC-SHA256/AES256-CBC-HMAC-SHA256.
* Pseudo Random Function (PRF) offload.

The README.md file also details the way in which various build options can be selected, under the section titled "## Intel&reg; Quickassist Technology OpenSSL\* Engine Build Options".  the majority of these options are enabled by default.  Does this answer your question?  If not, please could you specify which support algorithms you mean.

Regards,
Paul

**Comment by skomen** [comment]
I want to speed MD5 with QAT engine,is there any way to do it?

**Comment by stevelinsell** [comment]
MD5 support was not added to the QAT Engine due to the fact MD5 is: "considered cryptographically broken and unsuitable for further use".
The Quickassist hardware and driver are both capable of accelerating MD5 if required but you would need to add support for MD5 to the QAT Engine yourself. You would need to do this by implementing engine methods for digests for the NID: NID_md5.
An example of an md5 engine is available on the OpenSSL Blog at:
https://www.openssl.org/blog/blog/2015/11/23/engine-building-lesson-2-an-example-md5-engine/
As the digest methods structure is opaque on OpenSSL 1.1.0 you will need to use getters and setters in order to overide the methods rather than the example on the blog. 
If you do decide you need to implement md5 offload then there is some old code available from the OpenSSL 1.0.1 QAT Engine that would be a good starting point for porting across. Let me know if you would like me to point you to where you can obtain that code.

Kind Regards,

Steve.

**Comment by wujingcqu** [comment]
Is there anyway to make QAT engine support AES-GCM?

**Comment by stevelinsell** [comment]
Hi, @wujingcqu,

AES-GCM is a very efficient algorithm to perform on core using AES-NI rather than offload to an accelerator. Due to the cost of packaging up the request and unpackaging the response it only makes sense to consider offloading larger packets via the QAT Engine and even then the gains are small in comparison to offloading other operations. As such AES-GCM is currently not considered for addition to the QAT Engine. We plan to revisit that in the future post TLS 1.3 being ratified.

Kind Regards,

Steve.

**Comment by wujingcqu** [comment]
Hi, @stevelinsell ,
Thanks for your kind response. 

I am currently working with QAT for  academic research purposes. I want to evaluate/quantify the cost of various alternatives of hardware offloading for different algorithms, and possibly, in different operation modes, such as user space, kernel space, async/sync modes. 

If AES-GCM is not worthwhile for h/w offloading, it may not be included in QAT engine for production environment. I want to know if it is possible for me to add the support for AES-GCM support myself.
By the way, I need to add support for SHA1/256/, DES/3DES. Have you implemented these algorithms in the previous versions of QAT engine (1.0.1)? Can I use 1.0.1 as a start point to hack the 1.1.0 engine?

**Comment by stevelinsell** [comment]
Hi @wujingcqu,

You certainly could add AES-GCM support yourself. We did a very basic prototype a long while ago for the OpenSSL 1.0.1 QAT Engine and it did work ok functionally. Unfortunately the code was not of a quality that I can post it up here, but I can give advice if you do want to have a go.
In terms of the hashing functions SHA1 and SHA256, these were implemented in the OpenSSL 1.0.1 QAT Engine. It should be fairly straight forward to port these to work on the OpenSSL 1.1.0 QAT engine but I would recommend taking the synchronous code from the OpenSSL 1.0.1 QAT Engine as your base code and leave out all the asynchronous implementation. You can then convert it to work with the asynchronous features in OpenSSL 1.1.0 such as pausing the job afterwards, as asynchronous is much simpler in OpenSSL 1.1.0. The main things that will involve change are the fact OpenSSL made a lot of the structures opaque on OpenSSL 1.1.0. This will mean you need to change to use getters and setters where appropriate, and it substantially changes the way the EVP_MD methods are setup. You'd need to do something like the dasync engine does:
```
    _hidden_sha1_md = EVP_MD_meth_new(NID_sha1, NID_sha1WithRSAEncryption);
    if (_hidden_sha1_md == NULL
        || !EVP_MD_meth_set_result_size(_hidden_sha1_md, SHA_DIGEST_LENGTH)
        || !EVP_MD_meth_set_input_blocksize(_hidden_sha1_md, SHA_CBLOCK)
        || !EVP_MD_meth_set_app_datasize(_hidden_sha1_md,
                                         sizeof(EVP_MD *) + sizeof(SHA_CTX))
        || !EVP_MD_meth_set_flags(_hidden_sha1_md, EVP_MD_FLAG_DIGALGID_ABSENT)
        || !EVP_MD_meth_set_init(_hidden_sha1_md, dasync_sha1_init)
        || !EVP_MD_meth_set_update(_hidden_sha1_md, dasync_sha1_update)
        || !EVP_MD_meth_set_final(_hidden_sha1_md, dasync_sha1_final)) {
        EVP_MD_meth_free(_hidden_sha1_md);
        _hidden_sha1_md = NULL;
    }
```
With regard to the DES/3DES ciphers, again these were supported in the OpenSSL 1.0.1 QAT Engine but you'll have the same issues as the Digests above in converting them to work with OpenSSL 1.1.0.  You'll also find that you need to adjust the code to get them to work alongside the current chained ciphers implementations for AES128-SHA/AES256-SHA/AES128-SHA256/AES256-SHA256.
You can get a copy of the OpenSSL 1.0.1 QAT Engine [here](https://01.org/sites/default/files/page/libcrypto_shim_0.4.9-009_withdocumentation_0.zip).

Hope that helps,

Steve.

**Comment by stevelinsell** [comment]
Closing due to inactivity.
